### PR TITLE
espnet2 zeroth_korean recipe

### PR DESCRIPTION
**Pull request by hchung12**
This is an espnet2 recipe corresponding to the kaldi recipe(https://github.com/kaldi-asr/kaldi/tree/master/egs/zeroth_korean/s5) for  the Zeroth-Korean domain. 

Zeroth-Korean is a free Korean speech corpus, and it can be also downloaded from openslr:
http://www.openslr.org/40.

### CODE REVIEW

File: egs2/zeroth_korean/asr1/local/score.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#!/bin/bash
+# Copyright 2012  Johns Hopkins University (Author: Daniel Povey)
```
Comment (kamo-naoyuki) [inline egs2/zeroth_korean/asr1/local/score.sh:2]: Remove this script.

File: egs2/zeroth_korean/asr1/local/prepare_dict.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+
```
Comment (kamo-naoyuki) [inline egs2/zeroth_korean/asr1/local/prepare_dict.sh:2]: Remove this script.

File: egs2/zeroth_korean/asr1/local/format_lms.sh [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+
```
Comment (kamo-naoyuki) [inline egs2/zeroth_korean/asr1/local/format_lms.sh:2]: Remove this file.

File: egs2/zeroth_korean/asr1/local/update_segmentation.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
```
Comment (kamo-naoyuki) [inline egs2/zeroth_korean/asr1/local/update_segmentation.sh:2]: Remove this file.

File: egs2/zeroth_korean/asr1/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+- Git hash: `f2bd4912e32735487182c6f0cfddbf30e569a476`
+  - Commit date: `Thu Aug 6 08:31:46 2020 +0900`
+
+## asr_train_asr_transformer_fbank_pitch_bpe
+### WER
+
+|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
+|---|---|---|---|---|---|---|---|---|
+|inference_test_clean_inference_asr_model_valid.acc.best|457|6641|97.4|1.5|1.2|0.2|2.9|14.4|
+
```
Comment (kamo-naoyuki) [inline egs2/zeroth_korean/asr1/README.md:27]: Thank you for contributing! I didn't know this corpus and seems useful corpus.

This result looks good, is it expected result?

File: egs2/zeroth_korean/asr1/local/score.sh (hunk 1 above)
Comment (hchung12) [reply egs2/zeroth_korean/asr1/local/score.sh]: I removed the local/score.sh

File: egs2/zeroth_korean/asr1/local/prepare_dict.sh (hunk 2 above)
Comment (hchung12) [reply egs2/zeroth_korean/asr1/local/prepare_dict.sh]: I removed the local/prepare_dict.sh

File: egs2/zeroth_korean/asr1/local/update_segmentation.sh (hunk 4 above)
Comment (hchung12) [reply egs2/zeroth_korean/asr1/local/update_segmentation.sh]: I removed the update_segmentation.sh

File: egs2/zeroth_korean/asr1/README.md (hunk 5 above)
Comment (hchung12) [reply egs2/zeroth_korean/asr1/README.md:27]: I am happy to contribute to the ESPNet community. Yes, it is expected results. However, it should be noted that the model is trained using 4 GPUs.

File: egs2/zeroth_korean/asr1/local/format_lms.sh (hunk 3 above)
Comment (hchung12) [reply egs2/zeroth_korean/asr1/local/format_lms.sh]: I removed the local/format_lms.sh

File: egs2/zeroth_korean/asr1/README.md (hunk 5 above)
Comment (b-flo) [reply egs2/zeroth_korean/asr1/README.md:27]: Hi,

This model outperforms by a huge margin the best model from Kaldi (2.9 vs 9.45), is that expected? I'm not familiar with the recipe nor korean langage, sorry.
Also, looking at kaldi's RESULTS it seems scoring is done on a different number of words (6641 here vs 9253 in Kaldi, https://github.com/kaldi-asr/kaldi/blob/master/egs/zeroth_korean/s5/RESULTS), is there a reason?

Comment (hchung12) [reply egs2/zeroth_korean/asr1/README.md:27]: Yes, exactly.  
In the kaldi recipe,  a word  is first converted to a sequence of morpheme using the Morfessor.
So, the number of words (in actual the number of morphemes in kaldi) are different between the kaldi and espnet. 

Here are the first two lines of "text" of the test set for kaldi and espnet.
The parts marked in bold are different. You can see that words are more segmented in kaldi .
There are 43 words in kaldi but 32 words in espnet for the same lines.

--------------kaldi-----------------------------------
104_003_0019 지난해 삼 월 김 전 장관의 **동료 인** **장동 련** 홍익대 교수가 민간 **자문단 장으로** **위촉 되면서** 본격적인 **공모 와** 개발 작업에 들어갔다 
104_003_0193 그 **바람 에** **나 의** **몸 도** 겹쳐 쓰러지 며 한창 피어난 노란 동백꽃 속으로 폭 파묻혀 버렸다

--------------espnet-----------------------------------
104_003_0019 지난해 삼 월 김 전 장관의 **동료인** **장동련** 홍익대 교수가 민간 **자문단장으로** **위촉되면서** 본격적인 **공모와** 개발 작업에 들어갔다
104_003_0193 그 **바람에** **나의** **몸도** 겹쳐 쓰러지며 한창 피어난 노란 동백꽃 속으로 폭 파묻혀 버렸다

Comment (b-flo) [reply egs2/zeroth_korean/asr1/README.md:27]: Thanks for the explanation! In that case, I guess it would be better to explain the difference in RESULTS then?

Comment (hchung12) [reply egs2/zeroth_korean/asr1/README.md:27]: Okay, I will update README to explain why the RESULTS are different.

Comment (kamo-naoyuki) [reply egs2/zeroth_korean/asr1/README.md:27]: I'm  also  not familiar with korean lang. Just to make sure, why didn't you use Morfessor? Or, do you know the reason why Kaldi uses it?

Comment (hchung12) [reply egs2/zeroth_korean/asr1/README.md:27]: I don't think that morpheme is the best sub-word unit for Korean end-to-end ASR. 

Until Mike schuter proposed WPM in the paper "Japanese and korean voice search", morpheme was considered to the best sub-word unit to handle OOV as well as to consider contextual information in Korean ASR based on its linguistic characteristics such as agglutinative language. I think that kaldi uses the Morfessor for that reason.

However, as you know, in most end-to-end ASR framework, straightforward sub-word units having less linguistic prior such as grapheme, syllable or BPE are more common, and there is a report that end-to-end Korean ASR using syll-subword (6k) which is almost BPE shows the best performance on this zeroth-korean domain in the paper "https://arxiv.org/pdf/1910.11590.pdf". The paper reports 2.3% CER, and 2.4% WER.

Comment (kamo-naoyuki) [reply egs2/zeroth_korean/asr1/README.md:27]: Thank you, but I'm not still sure. 

Maybe people want to compare the result as fair as possible, so if  Morfessor is the standard way for this corpus, it's might be better to consider to apply it for scoring. 

This is just a problem how we define the words, isn't it? We can still use BPE after Morfessor segmentation.  Additionally, I understood that the paper did such processes as it said "We used the morphologically segmented text by using given morfessor model. ". Shouldn't we really use it?

Comment (hchung12) [reply egs2/zeroth_korean/asr1/README.md:27]: I agree that some people may want to compare the result using Morfessor. But, the problem is actually how to define WER.

In espnet recipe, WER means WER because it recovers words back from BPE sequences.

However, in kaldi, WER actually means morpheme error rate. 
Kaldi recipe uses morphemes  as if words after segmenting using Morfessor,  and measures error rate for recognized morphemes instead of decoding back to word sequence.

It causes the recognition difference. I think the WER metric of espnet seems more reasonable because it compares recognized results with the original transcription. I think it would be better that kaldi reports WER by comparing recognition results with original transcription. 

Anyway, I agree that it is useful to provide a recipe using Morfessor. So, I will update the recipe.

Thanks.

**Comment by sw005320** [comment]
@kamo-naoyuki, is it ready for @hchung12 to ask him to upload his model in our model zoo?

@hchung12, we have a new system to leverage the pre-trained model in https://github.com/espnet/espnet_model_zoo#register-your-model
Please consider it.

**Comment by hchung12** [comment]
@sw005320, I am ready to upload the model using word as output unit, and also finished to train another model using Morfessor to compare that of kaldi.  

I have a question for uploading models, I trained the models using 32 GPUs but I commit re-trained model using 4 GPUs to reproduce the results. In most cases, there is little difference. However, if the difference is not negligible, is it okay to upload the model using 32 GPUs ?

File: egs2/zeroth_korean/asr1/README.md (hunk 5 above)
Comment (hchung12) [reply egs2/zeroth_korean/asr1/README.md:27]: I update the recipe to use Morfessor as default and results. The WER is 2.3% which is even better than the previous result.

Thanks.

**Comment by kamo-naoyuki** [comment]
> is it okay to upload the model using 32 GPUs ?

It's okay.

File: egs2/zeroth_korean/asr1/README.md (hunk 5 above)
Comment (kamo-naoyuki) [reply egs2/zeroth_korean/asr1/README.md:27]: Thank you! This result seems reasonable.

File: egs2/zeroth_korean/asr1/local/update_segmentation.sh [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+dataDir=$1
+lmDir=$2
+
+exists(){
+	command -v "$1" >/dev/null 2>&1
+}
+
+# check morfessor installation 
+if ! exists morfessor; then
+	echo "You appear to not have Morfessor installed, either on your path."
```
Comment (kamo-naoyuki) [inline egs2/zeroth_korean/asr1/local/update_segmentation.sh:16]: Could you make tools/installers/install_morfessor.sh and add the path to path.sh?
Now path.sh is symbolic link, so you also need to copy it.

```sh
# in path.sh

export PATH=<morfessor>:$PATH
if ! command -v morfessor >/dev/null 2>&1; then
	echo "You appear to not have Morfessor installed, either on your path."
        echo "Run tools/installers/install_morfessor.sh"
	return 1
fi
```

Comment (hchung12) [reply egs2/zeroth_korean/asr1/local/update_segmentation.sh]: I updated the recipe.

File: tools/installers/install_morfessor.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  mkdir -p ./$dirname
+  git clone https://github.com/aalto-speech/morfessor.git morfessor ||
+    {
+      echo  >&2 "$0: Error git clone operation "
+      echo  >&2 "  Failed in cloning the github repository (https://github.com/aalto-speech/morfessor.git)"
+      exit
+    }
+fi
+
+echo >&2 "installation of MORFESSOR finished successfully"
```
Comment (kamo-naoyuki) [inline tools/installers/install_morfessor.sh:21]:
```suggestion
#

# Modified by Hoon Chung 2020 (ETRI) 
set -euo pipefail

echo "#### installing morfessor"
dirname=morfessor
rm -rf "${dirname}"
mkdir -p ./"${dirname}"
git clone https://github.com/aalto-speech/morfessor.git "${dirname}"
echo >&2 "installation of MORFESSOR finished successfully"
```

File: egs2/zeroth_korean/asr1/local/data.sh [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    local/download_and_untar.sh ${datadir}
+fi
+
+if [ $stage -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+  # format the data as Kaldi data directories
+  for part in train_data_01 test_data_01; do
+  	# use underscore-separated names in data directories.
+  	local/data_prep.sh ${datadir} ${part}
+  done
+fi
```
Comment (kamo-naoyuki) [inline egs2/zeroth_korean/asr1/local/data.sh:52]:
```suggestion
ndev_utt=220

log "$0 $*"
. utils/parse_options.sh

if [ $# -ne 0 ]; then
    log "Error: No positional arguments are required."
    exit 2
fi

. ./db.sh
. ./path.sh
. ./cmd.sh

train_set="train_nodev"
train_dev="train_dev"
test_set="test_clean"

if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
    log "stage 1: Data Download"
    mkdir -p "${ZEROTH_KOREAN}"
    local/download_and_untar.sh "${ZEROTH_KOREAN}"
fi

if [ $stage -le 1 ] && [ ${stop_stage} -ge 1 ]; then
  # format the data as Kaldi data directories
  for part in train_data_01 test_data_01; do
  	# use underscore-separated names in data directories.
  	local/data_prep.sh "${ZEROTH_KOREAN}" "${part}"
  done
fi  
```

**Comment by kamo-naoyuki** [comment]
I'll merge this PR. I'll commit small fix later.

**Comment by kamo-naoyuki** [comment]
@hchung12 Did you upload the model? (not mandatory)
If you have/ will have uploaded, please notify me.

**Comment by kamo-naoyuki** [comment]
Oh, I found that you are using fbank_pitch feats. It can't  used in model-zoo, so you don't need to upload it.
Why did you use pitch feats? It's efficient for korean?

**Comment by hchung12** [comment]
@kamo-naoyuki, there was some improvement when changing the feature from the default egs2/wsj setting. However, I don't think that fbank_pitch features are especially efficient for Korean.

I will do more experiments using raw feature next week and update it if the performance is similar or better.

**Comment by kamo-naoyuki** [comment]
Oh, thanks.  

FYI. batch_type=numel is more memory efficient.

https://github.com/espnet/espnet/blob/b4413f6259c49d2543db1e10417c08118a09d990/egs2/librispeech/asr1/conf/tuning/train_asr_transformer3.yaml#L1-L2

**Comment by kamo-naoyuki** [comment]
I also tested feats_type=raw with same setting, but I got worse result... pitch seems to improve much?

## asr_train_asr_transformer_raw_bpe
### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|inference_asr_model_valid.acc.ave/test_clean|457|9253|97.8|1.5|0.7|0.3|2.6|17.5|

### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|inference_asr_model_valid.acc.ave/test_clean|457|28084|98.7|0.6|0.7|0.3|1.6|17.5|

### TER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|inference_asr_model_valid.acc.ave/test_clean|457|14176|98.2|1.0|0.8|0.3|2.1|17.5|

**Comment by hchung12** [comment]
@kamo-naoyuki,  I am sorry for late updating the results.

I tuned hyper-paramters and got similar results using feats_type=raw and a different configuration setting with
batch_size: 64
accum_grad: 1

WER is lower but CER is better.
Do you think it is okay to replace the recipe with this one ?

## asr_train_asr_transformer5_raw_bpe
### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|test_clean|457|9253|97.8|1.4|0.9|0.2|2.4|17.3|

### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|test_clean|457|28084|98.6|0.5|0.9|0.2|1.6|17.3|

### TER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|test_clean|457|14176|98.2|0.9|0.9|0.2|2.0|17.3|

**Comment by kamo-naoyuki** [comment]
Oh, thanks.
Maybe, acceptable, I think. I think, we should compare using TER, originally, because this is the original result, bpe.

By the way, did you use averaged model? I recently updates some parts and now averaged model is the default.
And the scoring script was also fixed. Could you update your repository? You result doesn't have the information about decoding parameter, so.

**Comment by kamo-naoyuki** [comment]
Did you test lm? I recently tuned WSJ LM, https://github.com/espnet/espnet/blob/master/egs2/wsj/asr1/conf/tuning/train_lm_adam_layers4.yaml
(My previous lm configuration was also wrong, so you probably observed odd training result if you used it. )

If you have time, would you test it?

**Comment by hchung12** [comment]
@kamo-naoyuki, Ok, I will update my repository and re-run the experiment, and then PR in next week.
I will also test LM using the recipe.

**Comment by kamo-naoyuki** [comment]
Thanks. You don't need to re-run from scratch, it's enough to run from decoding stage, stage=10.
